Site Assembler - Add pattern categories

miksansegundo opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/wp-calypso/issues/72563 https://github.com/Automattic/wp-calypso/pull/74149 https://github.com/Automattic/wp-calypso/issues/74622

## Screenshots

|Closed state|Open state|
|---|---|
|<img width="1246" alt="Screenshot 2566-03-21 at 17 43 48" src="https://user-images.githubusercontent.com/1881481/226583599-1b8fdceb-23ac-4296-b2bb-887f2d10ef9d.png">|<img width="1246" alt="Screenshot 2566-03-21 at 17 44 04" src="https://user-images.githubusercontent.com/1881481/226583608-5f956d21-95d3-4071-86db-441050c59ed6.png">|

**Recording - Homepage > Add patterns** 

https://user-images.githubusercontent.com/1881481/226583142-8330dcda-09f1-4adb-a821-dc98d18cee7c.mov

## Proposed Changes

* Added the flag `pattern-assembler/categories`
* Updated the static patterns data (wireframe and posts patterns) to include categories
* Fetched categories from the endpoint `/block-patterns/categories`
* Show categories ordered as in the editor pattern inserter
* Click on a category to show the Pattern List Panel
* Click on the large preview to close the panel
* Updated copy to `Replace` a pattern
* Updated copy to the `Done` button to use `Save`
* Updated tooltip on pattern lists to show the pattern name 
* Pass the prop `pattern_category` in the tracks event `calypso_signup_pattern_assembler_pattern_select_click`
* Pass the prop `pattern_categories` in the tracks event `calypso_signup_pattern_assembler_pattern_select_done_click`

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Use the Calypso live link on the first comment
* Create a new site and click "Start designing" from the bottom of the design picker 
* Choose a header and a footer
* Create your homepage 
* Add patterns 
* Open categories
* Replace, order, and delete patterns
* Verify that the track events work as before and the `pattern_category` is added
* Continue and verify the patterns are added to a Home template on your site

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [X] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [X] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [X] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [X] Have you checked for TypeScript, React or other console errors?
- [X] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [X] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [X] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

miksansegundo committed:
Add categories with animation

miksansegundo committed:
fix

miksansegundo commented:
@davewhitley Thanks for your [feedback](https://github.com/Automattic/wp-calypso/pull/74060#issuecomment-1456830434) in the other PR. Can you test again in this PR? I updated the animation.

miksansegundo committed:
Code improvements from feedback

okmttdhr commented:
@miksansegundo Thank you for your PR! It's awesome that users are free to customize their content!

Here are some thoughts from me as a new dev;

1. (just an idea) I would like to see a preview that scrolls to the corresponding UI when the pattern item is moused over or clicked (Since I felt that if I added several patterns, it would be difficult to grasp what my site looked like).

<img width="1422" alt="Screen Shot 2023-03-08 at 14 42 25" src="https://user-images.githubusercontent.com/5287479/223629467-816d2b29-5e5f-473a-94da-ab412d48a280.png">

2. Related to keyboard accessibility mentioned in pdtkmj-1eO-p2#comment-2276, I thought it would be more intuitive for the user if, after selecting a pattern category with the `Enter` key, the user presses the tab to focus on the pattern. We have to press the tab several times, as shown below now;

https://user-images.githubusercontent.com/5287479/223629730-0ee73c4b-bf28-4d92-ac08-05979949784a.mov


3. Last but not least, `<` button and `action bar` addressed in pdtkmj-1eO-p2#comment-2274 will improve UX here, I believe!

miksansegundo committed:
Code improvements

miksansegundo committed:
Code improvements

miksansegundo commented on client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/hooks/use-categories-order.ts:
```diff
@@ -0,0 +1,45 @@
+import { useMemo } from 'react';
+import type { Category } from '../types';
+
+const patternCategoriesOrder = [
```
I copied the implementation from the editor because I find it "fine" https://github.com/WordPress/gutenberg/pull/47760/files

arthur791004 replied:
Got it, thanks for the explanation!

miksansegundo commented on client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/hooks/use-patterns-map-by-category.ts:
```diff
@@ -0,0 +1,23 @@
+			{}
+		) as { [ key: string ]: Pattern[] };
+
+		patterns.forEach( ( pattern ) => {
```
I implemented an optimized map version to iterate the patterns only once.

miksansegundo commented on client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/index.tsx:
```diff
@@ -60,6 +63,13 @@ const PatternAssembler: Step = ( { navigation, flow, stepName } ) => {
 	const allPatterns = useAllPatterns();
+	const sectionPatterns = useSectionPatterns();
+
+	// Fetching the categories so they are ready when ScreenCategoryList loads
```
The logic to fetch categories and create `sectionsMapByCategory` is here to avoid doing it in `ScreenCategoryList` because that causes a noticeable delay and the screen is blank for a moment. Any better ideas?

arthur791004 replied:
Yup, I know. It's fine to simply keep it here since we will need to fetch all patterns in the future. Also, we did fetching the rendered patterns at the beginning. We can iterate it later when we have a better ideas

miksansegundo committed:
Fix bug onSelect

miksansegundo commented on client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-list-panel.tsx:
```diff
@@ -0,0 +1,59 @@
+						layoutScroll
+						initial={ { opacity: 0, transform: 'translateX(0)' } }
+						animate={ { opacity: 1, transform: 'translateX(348px)' } }
+						transition={ {
```
@davewhitley here are the params for the animation. Happy to try other values.

miksansegundo commented on client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/patterns-data.ts:
```diff
@@ -202,359 +168,287 @@ const useFooterPatterns = () => {
 
-	const sectionPatterns: Pattern[] = useMemo( () => {
-		let patterns = [
+	const sectionPatterns: Pattern[] = useMemo(
```
We'll replace the list of sections with a request to `dotcompatterns` in a following PR

miksansegundo commented on client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/screen-header.tsx:
```diff
@@ -26,13 +28,25 @@ const ScreenHeader = ( { selectedPattern, onSelect, onBack, onDoneClick }: Props
 				<PatternSelector
 					patterns={ patterns }
-					onSelect={ onSelect }
+					onSelect={ ( selectedPattern ) => onSelect( 'header', selectedPattern, 'header' ) }
```
I don't like to repeat the param `header` here, but one is the type and the other is the category. However, it is better than using `null` and then managing it in `onSelect`.

davewhitley commented:
I don't think the category text is necessary right now.

It seems like the animation can't be improved unless we change how the iframe patterns are loaded. The categories with only 2 or 3 patterns load much smoother with the sidebar animation. Thinking more about it, perhaps we shouldn't do an animation for the sidebar panel (for now).

autumnfjeld commented:
> I would like to see a preview that scrolls to the corresponding UI when the pattern item is moused over or clicked (Since I felt that if I added several patterns, it would be difficult to grasp what my site looked like).

Very cool idea!   @okmttdhr could you create an issue for this and put it in the backlog?

miksansegundo committed:
Fix types

miksansegundo committed:
Fix types

miksansegundo committed:
Adjust sidebar styles

miksansegundo committed:
Optimize panel animation

miksansegundo committed:
Fix to show shadow

miksansegundo commented:
> It seems like the animation can't be improved unless we change how the iframe patterns are loaded. The categories with only 2 or 3 patterns load much smoother with the sidebar animation. 

The list of iframe patterns is heavy for sure. I have removed the opacity animation and added some optimizations. Please test it again. Does it look better to you? I think it can be good enough for a first iteration. @davewhitley 

> Thinking more about it, perhaps we shouldn't do an animation for the sidebar panel (for now).
We can still try doing the animation with only CSS. I'll continue working on that later after getting feedback on this new version.

miksansegundo committed:
Clean onBack

miksansegundo committed:
Adjust rendering patterns timeout to 200ms

okmttdhr commented:
> > I would like to see a preview that scrolls to the corresponding UI when the pattern item is moused over or clicked (Since I felt that if I added several patterns, it would be difficult to grasp what my site looked like).
> 
> Very cool idea! @okmttdhr could you create an issue for this and put it in the backlog? We can consider this not needed for the Call for Testing or shipping to users as part of categories, but It think it would be a very nice feature to add!

Created https://github.com/Automattic/wp-calypso/issues/74246 👍 Thank you!

arthur791004 commented:
The diff looks weird 😂 You might need to rebase it onto the base branch (?)

arthur791004 commented on client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-selector.tsx:
```diff
@@ -1,54 +1,35 @@
-						emptyPatternText={ emptyPatternText }
-						activeClassName="pattern-selector__block-list--selected-pattern"
-						onSelect={ onSelect }
+						{ ...{
```
I guess nothing changed here 😅

arthur791004 commented on client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/screen-footer.tsx:
```diff
@@ -38,8 +38,10 @@ const ScreenFooter = ( { selectedPattern, onSelect, onBack, onDoneClick }: Props
 				<NavigatorBackButton
 					as={ Button }
 					className="pattern-assembler__button"
+					onClick={ () => {
```
Same

arthur791004 commented on client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/screen-header.tsx:
```diff
@@ -38,8 +37,10 @@ const ScreenHeader = ( { selectedPattern, onSelect, onBack, onDoneClick }: Props
 				<NavigatorBackButton
 					as={ Button }
 					className="pattern-assembler__button"
+					onClick={ () => {
```
Same

arthur791004 commented on client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/screen-pattern-list.tsx:
```diff
@@ -13,27 +14,30 @@ interface Props {
 					className="pattern-assembler__button"
-					primary
 					onClick={ onDoneClick }
+					primary
```
Same

arthur791004 commented on client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-list-panel.tsx:
```diff
@@ -0,0 +1,58 @@
+	const categoryPatterns = categorySelected ? sectionPatternsMapByCategory[ categorySelected ] : [];
+
+	return (
+		<AnimateMotion featureName="domAnimation">
```
In this way, the framer-motion will be bundled with the PA step and that's why we need to use `AsyncLoad`. It would be better to apply the `AsycLoad` here to keep the size of the PA step smaller. WDYT?

![image](https://user-images.githubusercontent.com/13596067/224246669-a540790e-3611-4a48-a3d2-3eb918db9279.png)

miksansegundo replied:
Good catch. Because the `PanelListPanel` is only in `CategoryScreenList`, I'll use `AsycLoad` to load `CategoryScreenList` when the flag is enabled. I think that works too, right?

```
<NavigatorScreen path="/homepage/patterns">
  { isEnabled( 'pattern-assembler/categories' ) ? (
    <AsyncLoad
      require="./screen-category-list"
```

arthur791004 replied:
Yes, it makes sense 👍

arthur791004 commented on client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/index.tsx:
```diff
@@ -243,13 +253,29 @@ const PatternAssembler: Step = ( { navigation, flow, stepName } ) => {
-				patternId: selectedPattern?.id || 0,
-				patternName: selectedPattern?.name || '',
+				patternType: selectedPattern.category?.name || type,
+				patternId: selectedPattern.id,
```
We use the `null` value to track deselecting header and footer. If we wrap the event inside the check, we're not able to know whether the user removes the header and footer. Any thougths?

miksansegundo replied:
In tracks, other ways exist to know when the header and footer are not selected. In `calypso_signup_pattern_assembler_continue_click` with the prop `pattern_types` or in the event `calypso_signup_pattern_assembler_pattern_final_select` with the prop `pattern_category`.


I think `calypso_signup_pattern_assembler_pattern_select_click` makes more sense to track patterns selected rather than patterns not selected.

arthur791004 replied:
Okay, that also makes sense to me 🙂

arthur791004 commented on client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/screen-pattern-list.tsx:
```diff
@@ -13,27 +14,30 @@ interface Props {
 	return (
 		<>
-			<NavigatorHeader title={ translate( 'Add patterns' ) } />
+			{ ! isEnabled( 'pattern-assembler/categories' ) && (
```
I guess we don't need to check this as we won't use the `ScreenPatternList` when the flag is enabled.

arthur791004 commented on client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/screen-category-list.tsx:
```diff
@@ -0,0 +1,129 @@
+}: Props ) => {
+	const translate = useTranslate();
+	const [ categorySelected, setCategory ] = useState< string | null >( null );
+	const [ openPatternList, setOpenPatternList ] = useState< boolean | null >( null );
```
Looks like this state depends on the `categorySelected`. We can simply derive the value from it. Does it make sense?

miksansegundo replied:
The idea here is to have two different states to control the animation independently of the loading of patterns. Thanks to this we can delay the rendering of patterns `200ms`.

arthur791004 replied:
We can delay setting the category if it is never selected as we use `category` to delay the rendering instead of `openPatternList`. I'm okay with the current way 🙂

arthur791004 commented on client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/screen-category-list.tsx:
```diff
@@ -0,0 +1,129 @@
+	wrapperRef,
+}: Props ) => {
+	const translate = useTranslate();
+	const [ categorySelected, setCategory ] = useState< string | null >( null );
```
[nit] Could we align the naming? e.g `selectedCategory` and `setSelectedCategory`. It also aligns with `selectedPattern`.

miksansegundo replied:
Good catch. I missed that.

arthur791004 commented on client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/screen-category-list.tsx:
```diff
@@ -0,0 +1,129 @@
+	const sectionPatterns = useSectionPatterns();
+	const categoriesInOrder = useCategoriesOrder( categories );
+
+	const categoriesWithPatterns = useMemo( () => {
```
Maybe we can simply render `null` if the category doesn't have any pattern, and then we're able to get rid of this memo. WDYT?

miksansegundo replied:
Sure, that's better.

arthur791004 commented on client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-list-panel.scss:
```diff
@@ -0,0 +1,17 @@
+	position: absolute;
+	z-index: 1;
+	top: -60px;
+	left: 0;
```
We have to handle the right-to-left direction. It would be better to do it in the follow-up PR as there are some issues there 😓

miksansegundo replied:
Okay! https://github.com/Automattic/wp-calypso/issues/74320

miksansegundo committed:
Address feedback

arthur791004 commented:
BTW, I found there is a horizontal scrollbar when you disabled the feature flag. 

<img width="300" src="https://user-images.githubusercontent.com/13596067/224266134-6b2e7e4f-aa55-417d-b943-1e88012bb8e5.png" />

arthur791004 commented on client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/screen-category-list.tsx:
```diff
@@ -0,0 +1,130 @@
+	};
+
+	return (
+		<div className="screen-container" { ...useFocusOutside( handleFocusOutside ) }>
```
We need to add `tabIndex={ -1 }` to make `useFocusOutside` work as expected. Now if you click on the screen container (not on the category), the pattern list panel will be closed and it's not user-friendly. However, there is another issue when you click on the pattern list panel (not on the pattern) since it's not a child of the screen container. Let's tackle it in the follow-up PR

miksansegundo replied:
`useFocusOutside` doesn't work on Safari, and when it works it doesn't do what we want, which is "Click on the large preview to close Pattern List". That's why I'm adding back the previous solution that works perfectly.

miksansegundo committed:
Fix click on large preview to close Pattern List

miksansegundo committed:
Clean event listener on unmount

miksansegundo commented on client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/screen-category-list.tsx:
```diff
@@ -0,0 +1,136 @@
+	};
+	useEffect( () => {
+		wrapperRef?.current?.addEventListener( 'click', handleFocusOutside );
+		return () => wrapperRef?.current?.removeEventListener( 'click', handleFocusOutside );
```
Because the screen unmounts every time it is shown, we must remove the listener, but it's not doing it. I think I must use a ref for `handleFocusOutside`. 

I don't know why the event is not added multiple times, anyways that is good news.

<img width="495" alt="Screenshot 2566-03-10 at 16 27 00" src="https://user-images.githubusercontent.com/1881481/224278672-48269670-42c2-458e-b55f-5f49d4eb1e2e.png">

miksansegundo replied:
It's working without using a ref.
<img width="496" alt="Screenshot 2566-03-10 at 16 38 57" src="https://user-images.githubusercontent.com/1881481/224280766-ef8cd68e-e56e-43f4-b3f7-0ea13fc549af.png">

I had to close and open Chome Dev tools again to see the event removed. In Safari, the list of events is updated immediately tho.

miksansegundo committed:
Remove panel animation

miksansegundo committed:
Fix scroll on footer

miksansegundo committed:
Revert

miksansegundo committed:
Fix category selected and tracks event

miksansegundo committed:
Style the category list with more consistency

miksansegundo commented:
> BTW, I found there is a horizontal scrollbar when you disabled the feature flag.

Done ✅

I have also removed the panel animation. I'll address any pending feedback without adding more features.

arthur791004 commented on client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/pattern-list-panel.tsx:
```diff
@@ -0,0 +1,40 @@
+	categories,
+	selectedCategory,
+}: PatternListPanelProps ) => {
+	const sectionPatternsMapByCategory = useSectionPatternsMapByCategory( patterns, categories );
```
Looks like we can simply pass `PatternsMapByCategory` to this component. WDYT?

miksansegundo replied:
Nice 👍

arthur791004 commented on client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/index.tsx:
```diff
@@ -57,6 +60,13 @@ const PatternAssembler: Step = ( { navigation, flow, stepName } ) => {
+
+	// Fetching the categories so they are ready when ScreenCategoryList loads
+	const categoriesQuery = usePatternCategories( site?.ID );
+	const categories = ( categoriesQuery?.data || [] ) as Category[];
```
[nit] We can let `usePatternCategories` return the categories directly. But it's optional and not a blocker.

miksansegundo replied:
I tried that and ran into issues because of the typescript linter. I'll happily update it if you can show me an example.

arthur791004 replied:
Maybe https://github.com/Automattic/wp-calypso/compare/feature/site-assembler-categories...feature/site-assembler-categories-example?expand=1

miksansegundo replied:
Thanks! I'll try this later.

miksansegundo replied:
I'm looking into this, but the link is not working. Can you share it again, please?

arthur791004 replied:
https://github.com/Automattic/wp-calypso/commit/fdf667ab0b7c6598daea6ce5279ca482ff20a2c0

arthur791004 commented on client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/screen-footer.tsx:
```diff
@@ -38,8 +37,8 @@ const ScreenFooter = ( { selectedPattern, onSelect, onBack, onDoneClick }: Props
 					className="pattern-assembler__button"
-					primary
 					onClick={ onDoneClick }
+					primary
```
Looks like we don't need this change 😅

arthur791004 commented on client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/screen-header.tsx:
```diff
@@ -38,8 +37,8 @@ const ScreenHeader = ( { selectedPattern, onSelect, onBack, onDoneClick }: Props
 					className="pattern-assembler__button"
-					primary
 					onClick={ onDoneClick }
+					primary
```
Looks like we don't need this change 😅

arthur791004 reviewed (approved):
Tested and looks good to me. Leave some minor feedback but it's optional 🙂

miksansegundo committed:
Code improvements

miksansegundo commented on client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/screen-category-list.tsx:
```diff
@@ -0,0 +1,131 @@
+					{ translate( 'Done' ) }
+				</NavigatorBackButton>
+			</div>
+			{ createPortal(
```
I'll remove the portal and the CSS related to the animation because it's not required now that we don't have animation.

miksansegundo commented:
I've addressed the feedback. Thanks @arthur791004. I also removed the portal and CSS related to the animation because it's not required

miksansegundo committed:
Fix type

miksansegundo committed:
Update button copy to use Save

miksansegundo committed:
Fix category name onSelect

miksansegundo committed:
Add stylelint-disable-next-line

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7879722

Thank you @miksansegundo for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7879722).
